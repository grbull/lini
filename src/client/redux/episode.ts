/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type EpisodeState = {
  isLoading: boolean;
  error?: string;
  data?: EpisodeDto;
};

const initialState: EpisodeState = {
  isLoading: false,
};

export const get = createAsyncThunk<EpisodeDto, number>('episode/get', (id) =>
  api.episode.get(id)
);

export const episode = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const episodeActions = { ...episode.actions, get };
