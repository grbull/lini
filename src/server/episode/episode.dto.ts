import { ShowDto } from '../show/show.dto';

export class EpisodeDto {
  id!: string;
  name!: string;
  season!: number;
  number!: number;
  type!: string;
  airdate?: string;
  airtime?: string;
  airstamp?: string;
  runtime?: number;
  imageMedium?: string;
  imageOriginal?: string;
  summary?: string;
  show!: ShowDto;
}

export class ScheduleDto {
  future!: EpisodeDto[];
  past!: EpisodeDto[];
}
