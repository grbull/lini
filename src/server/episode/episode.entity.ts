import { Column, Entity, ManyToOne } from 'typeorm';

import { ShowEntity } from '../show/show.entity';

@Entity('episode')
export class EpisodeEntity {
  @Column('bigint', { primary: true })
  id!: string;

  @Column('text')
  name!: string;

  @Column('integer')
  season!: number;

  @Column('integer')
  number!: number;

  @Column('text')
  type!: string;

  @Column('date', { nullable: true })
  airdate!: string | null;

  @Column('time', { nullable: true })
  airtime!: string | null;

  @Column('timestamptz', { nullable: true })
  airstamp!: string | null;

  @Column('integer', { nullable: true })
  runtime!: number | null;

  @Column('text', { nullable: true })
  imageMedium!: string | null;

  @Column('text', { nullable: true })
  imageOriginal!: string | null;

  @Column('text', { nullable: true })
  summary!: string | null;

  @ManyToOne(() => ShowEntity, (show) => show.episodes)
  show!: number;
}
