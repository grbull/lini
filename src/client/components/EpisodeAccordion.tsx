import cn from 'classnames';
import React, { ReactElement, useState } from 'react';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { EpisodeAccordionItem } from './EpisodeAccordionItem';

interface Props {
  episodes: Omit<EpisodeDto, 'show'>[];
}

export function EpisodeAccordion({ episodes }: Props): ReactElement {
  const [expanded, setExpanded] = useState(0);

  const seasons = episodes
    .reduce((seasons: number[], episode) => {
      if (seasons.find((season) => episode.season === season)) {
        return seasons;
      }
      return [...seasons, episode.season];
    }, [])
    .sort((a, b) => a - b);

  return (
    <>
      <h2 className={cn('text-xl', 'mb-2')}>Episode List</h2>

      {seasons.map((season) => (
        <EpisodeAccordionItem
          episodes={episodes.filter((episode) => episode.season === season)}
          isExpanded={expanded === season}
          key={season}
          season={season}
          setExpanded={setExpanded}
        />
      ))}
    </>
  );
}
