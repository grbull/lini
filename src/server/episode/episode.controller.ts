import {
  Controller,
  Get,
  Param,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';

import { User } from '../common/decorators/user.decorator';
import { AuthGuard } from '../common/guards/auth.guard';
import { LoggingInterceptor } from '../common/interceptors/logging.interceptor';
import { UserEntity } from '../user/user.entity';
import { EpisodeEntity } from './episode.entity';
import { EpisodeService } from './episode.service';

@Controller('episode')
@UseGuards(AuthGuard)
@UseInterceptors(LoggingInterceptor)
export class EpisodeController {
  constructor(private readonly episodeService: EpisodeService) {}

  @Get('schedule')
  public async schedule(
    @User() user: UserEntity
  ): Promise<{ future: EpisodeEntity[]; past: EpisodeEntity[] }> {
    const future = await this.episodeService.getFuture(user);
    const past = await this.episodeService.getPast(user);
    return { future, past };
  }

  @Get(':id')
  public getOne(@Param('id') id: number): Promise<EpisodeEntity> {
    return this.episodeService.getOne(id);
  }
}
