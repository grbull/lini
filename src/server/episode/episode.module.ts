import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { LoggerModule } from '../logger/logger.module';
import { SubscriptionModule } from '../subscription/subscription.module';
import { EpisodeController } from './episode.controller';
import { EpisodeEntity } from './episode.entity';
import { EpisodeService } from './episode.service';

@Module({
  imports: [
    LoggerModule,
    TypeOrmModule.forFeature([EpisodeEntity]),
    SubscriptionModule,
  ],
  providers: [EpisodeService],
  controllers: [EpisodeController],
  exports: [EpisodeService],
})
export class EpisodeModule {}
