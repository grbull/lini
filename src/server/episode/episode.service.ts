import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Between, Repository } from 'typeorm';

import { LoggerService } from '../logger/logger.service';
import { SubscriptionService } from '../subscription/subscription.service';
import { TvMazeEpisode } from '../tvmaze/tvmaze.types';
import { tvMazeEpisodeToEntity } from '../tvmaze/tvmaze.utils';
import { UserEntity } from '../user/user.entity';
import { EpisodeEntity } from './episode.entity';

@Injectable()
export class EpisodeService {
  constructor(
    private readonly loggerService: LoggerService,
    @InjectRepository(EpisodeEntity)
    private readonly episodeRepository: Repository<EpisodeEntity>,
    private readonly subscriptionService: SubscriptionService
  ) {
    this.loggerService.setContext('EpisodeService');
  }

  public async getOne(id: number): Promise<EpisodeEntity> {
    try {
      return await this.episodeRepository
        .createQueryBuilder('episode')
        .leftJoinAndSelect('episode.show', 'show')
        .where('episode.id = :id', { id })
        .getOneOrFail();
    } catch {
      throw new NotFoundException();
    }
  }

  // public getByShow(id: number): Promise<EpisodeEntity[]> {
  //   return this.episodeRepository
  //     .createQueryBuilder('episode')
  //     .where('episode.show.id = :id', { id })
  //     .orderBy('episode.airdate', 'ASC')
  //     .getMany();
  // }

  // public async getUpcoming(user: UserEntity): Promise<EpisodeEntity[]> {
  //   const subscriptions = await this.subscriptionService.getAll(user);
  //   let upcomingEpisodes: EpisodeEntity[] = [];

  //   const oneMonthAhead = new Date();
  //   oneMonthAhead.setMonth(new Date().getMonth() + 1);

  //   for (const key of subscriptions) {
  //     const episodes = await this.episodeRepository.find({
  //       where: {
  //         // between now and 1 month ahead
  //         // airstamp: Raw(
  //         //   (alias) => `${alias} < '${oneMonthAhead.toISOString()}'`
  //         // ),
  //         airstamp: Between(new Date(), oneMonthAhead),
  //         show: key.show,
  //       },
  //     });
  //     console.log(episodes);
  //     if (episodes) {
  //       upcomingEpisodes = [...upcomingEpisodes, ...episodes];
  //     }
  //   }
  //   return upcomingEpisodes;
  // }

  public async getFuture(user: UserEntity): Promise<EpisodeEntity[]> {
    const subscriptions = await this.subscriptionService.getAll(user);

    const oneMonthAhead = new Date();
    oneMonthAhead.setMonth(new Date().getMonth() + 1);

    const findOptions = subscriptions.map((subscription) => {
      return {
        airstamp: Between(new Date(), oneMonthAhead),
        show: subscription.show,
      };
    });

    return this.episodeRepository.find({
      where: findOptions,
      order: {
        airstamp: 'ASC',
        number: 'ASC',
      },
      relations: ['show'],
    });
  }

  public async getPast(user: UserEntity): Promise<EpisodeEntity[]> {
    const subscriptions = await this.subscriptionService.getAll(user);

    const oneMonthBehind = new Date();
    oneMonthBehind.setMonth(new Date().getMonth() - 1);

    const findOptions = subscriptions.map((subscription) => {
      return {
        airstamp: Between(oneMonthBehind, new Date()),
        show: subscription.show,
      };
    });

    return this.episodeRepository.find({
      where: findOptions,
      order: {
        airstamp: 'DESC',
        number: 'DESC',
      },
      relations: ['show'],
    });
  }

  // The belowe is used for seeding and updating
  public createOrUpdateMany(
    episodes: TvMazeEpisode[],
    showId: number
  ): Promise<EpisodeEntity[]> {
    const episodeEntities = episodes.map((episode) =>
      tvMazeEpisodeToEntity(episode, showId)
    );

    return this.episodeRepository.save(episodeEntities, { chunk: 1000 });
  }

  // used by notification service
  public findByTime(
    showId: number,
    startTime: Date,
    endTime: Date
  ): Promise<EpisodeEntity[]> {
    return this.episodeRepository.find({
      where: { show: showId, airstamp: Between(startTime, endTime) },
    });
  }
}
