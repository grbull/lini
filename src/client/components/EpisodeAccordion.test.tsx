/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import { render } from '@testing-library/react';
import React from 'react';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { EpisodeAccordion } from './EpisodeAccordion';

const episodes: Omit<EpisodeDto, 'show'>[] = [
  {
    id: '1627015',
    name: 'Red Moon',
    season: 1,
    number: 1,
    type: 'regular',
    airdate: '2019-11-01',
    airtime: null,
    airstamp: '2019-11-01T12:00:00.000Z',
    runtime: 65,
    imageMedium: '/uploads/images/medium_landscape/218/545490.jpg',
    imageOriginal: '/uploads/images/original_untouched/218/545490.jpg',
    summary:
      'NASA is in crisis as the Soviets land the first man on the moon in 1969, the beginning of an alternate history.',
  },
  {
    id: '1727898',
    name: 'He Built the Saturn V',
    season: 1,
    number: 2,
    type: 'regular',
    airdate: '2019-11-01',
    airtime: null,
    airstamp: '2019-11-01T12:00:00.000Z',
    runtime: 62,
    imageMedium: '/uploads/images/medium_landscape/218/545491.jpg',
    imageOriginal: '/uploads/images/original_untouched/218/545491.jpg',
    summary:
      "Director von Braun opposes President Nixon's directive, with dire consequences.",
  },
  {
    id: '1972621',
    name: 'Every Little Thing',
    season: 2,
    number: 1,
    type: 'regular',
    airdate: '2021-02-19',
    airtime: null,
    airstamp: '2021-02-19T12:00:00.000Z',
    runtime: 56,
    imageMedium: '/uploads/images/medium_landscape/284/710545.jpg',
    imageOriginal: '/uploads/images/original_untouched/284/710545.jpg',
    summary:
      'Nearly a decade later, technology and lunar exploration have taken huge stridesâ€”but a solar storm threatens the astronauts on Jamestown.',
  },
];

describe('EpisodeAccordion Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = render(<EpisodeAccordion episodes={episodes} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render two episode accordion items', async () => {
    const { findAllByRole } = render(<EpisodeAccordion episodes={episodes} />);
    expect(await findAllByRole('button')).toHaveLength(2);
  });
});
