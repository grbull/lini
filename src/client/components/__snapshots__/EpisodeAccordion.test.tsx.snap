// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeAccordion Component matches the snapshot 1`] = `
<DocumentFragment>
  <h2
    class="text-xl mb-2"
  >
    Episode List
  </h2>
  <div
    class="py-1"
    role="button"
    tabindex="1"
  >
    <div
      class="flex justify-between items-center"
    >
      <span>
        Season 1
      </span>
      <span>
        2 Episodes
      </span>
    </div>
  </div>
  <div
    class="py-1"
    role="button"
    tabindex="2"
  >
    <div
      class="flex justify-between items-center"
    >
      <span>
        Season 2
      </span>
      <span>
        1 Episodes
      </span>
    </div>
  </div>
</DocumentFragment>
`;
